File: 99-Continue/166/ZigZagTraversal.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.data = data
        self.left = self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def buildTree(self,arr:list[str])->Node:
        if len(arr)==0:
            return self.root
            
        self.root = Node(int(arr[0]))
        queue = deque([self.root])
        i = 1
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr) and arr[i]!="null":
                current.left = Node(int(arr[i]))
                queue.append(current.left)
            i+=1
            if i<len(arr) and arr[i]!="null":
                current.right = Node(int(arr[i]))
                queue.append(current.right)
            i+=1
        return self.root
    
    def zigZagPrintTree(self)->Node:
        if self.root is None:
            return self.root
        
        queue = deque([self.root])
        print(self.root.data,end="")
        while queue:
            size = len(queue)
            print()
            while size>0:
                current = queue.popleft()
                if current.right is not None:
                    if current.left is not None:
                        queue.append(current.left)
                    print(current.right.data,end=" ")
                if current.left is not None:
                    if current.right is not None:
                        queue.append(current.right)
                    print(current.left.data,end=" ")
                size-=1
            print()
            size0 = len(queue)
            while size0>0:
                current = queue.popleft()
                if current.left is not None:
                    queue.append(current.left)
                    print(current.left.data,end=" ")
                if current.right is not None:
                    queue.append(current.right)
                    print(current.right.data,end=" ")
                size0-=1
                

def main():
    int(input())
    s = input().split()
    tree = BinaryTree()
    tree.buildTree(s)
    tree.zigZagPrintTree()
    
if __name__=="__main__":
    main()
            
            
                
        
    
    
        

File: 99-Continue/166/ZigZagTraversal.java

import java.util.*;

public class ZigZagTraversal {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();
        String s = sc.nextLine();
        String[] strarr = s.split(" ");
        BinaryTree tree = new BinaryTree();
        tree.buildTree(strarr);
        tree.printTree();
        sc.close();
    }
}

class Node{
    int data;
    Node left,right;
    Node(int data){
        this.data = data;
        this.left = this.right = null;
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root = null;
    }
    
    Node buildTree(String[] arr){
        if(arr.length==0){
            return root;
        }
        root = new Node(Integer.parseInt(arr[0]));
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        int i =1;
        while(!queue.isEmpty() && i<arr.length){
            Node current = queue.poll();
            if(i<arr.length && !arr[i].equals("null")){
                current.left = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.left);
            }
            i++;
            if(i<arr.length && !arr[i].equals("null")){
                current.right = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.right);
            }
            i++;
        }
        return root;
    }
    
    void printTree(){
        if(root==null){
            return;
        }
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        System.out.print(root.data);
        while(!queue.isEmpty()){
            System.out.println();
            int size = queue.size();
                while(size-->0){
                Node current = queue.poll();
               
                if(current.right!=null){
                    if(current.left!=null){
                        queue.offer(current.left);
                    }
                    System.out.print(current.right.data+" ");
                }
                 if(current.left!=null){
                    if(current.right!=null){
                        queue.offer(current.right);
                    }
                    System.out.print(current.left.data+" ");
                }
            }
            System.out.println();
            int size0 = queue.size();
                while(size0-->0){
                Node current = queue.poll();
                 if(current.left!=null){
                    queue.offer(current.left);
                    System.out.print(current.left.data+" ");
                }
                
                if(current.right!=null){
                    queue.offer(current.right);
                    System.out.print(current.right.data+" ");
                }
            }
           
        }
    }
}